### PR TITLE
[Issue 2566] Add deposit event validation

### PR DESCRIPTION
**Pull request by mbaxter**
## PR Description
Add validation to ensure that eth1 deposit events are ordered and contiguous:
* Add validation to `DepositsFromBlockEvent` to make sure the deposit events included are contiguous (there are no gaps / missing deposits).
* Add `ValidatingEth1EventsPublisher` which tracks the latest deposit index processed and ensure that the next deposit event received starts at the expected index
* When replaying deposits in `DepositStorage`:
  * Make sure we start from the first deposit at index 0
  * Track the last deposit processed to ensure that no deposits are missed
* Add a fatal log message if an unhandled `InvalidDepositEventsException` is caught.

## Fixed Issue(s)
Part of #2566 

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: pow/src/main/java/tech/pegasys/teku/pow/ValidatingEth1EventsPublisher.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    delegate.onDepositsFromBlock(event);
+  }
+
+  private void assertDepositEventIsValid(final DepositsFromBlockEvent event) {
+    final Optional<UInt64> expectedFirstDepositIndex = lastPublishedDeposit.map(UInt64::increment);
+    final boolean depositIndexIsValid =
+        expectedFirstDepositIndex
+            .map(expected -> expected.equals(event.getFirstDepositIndex()))
+            .orElse(true);
```
Comment (ajsutton) [inline pow/src/main/java/tech/pegasys/teku/pow/ValidatingEth1EventsPublisher.java]: I suspect this misses the case where we miss the very first deposit.  We won't have replayed anything from disk so the last deposit index will be unset and we should then expect the first deposit to have index 0.

File: pow/src/main/java/tech/pegasys/teku/pow/exception/InvalidDepositEventsException.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+public class InvalidDepositEventsException extends IllegalArgumentException {
+
+  public InvalidDepositEventsException(final String s) {
+    super(s);
+  }
+
+  public static InvalidDepositEventsException expectedDepositAtIndex(UInt64 expectedIndex) {
+    return new InvalidDepositEventsException("Expected next deposit at index " + expectedIndex);
+  }
```
Comment (ajsutton) [inline pow/src/main/java/tech/pegasys/teku/pow/exception/InvalidDepositEventsException.java]: Should we include the actual deposit index we got in the message?

**Review by ajsutton — APPROVED** [review]
LGTM. Just worth checking that it detects if the very first deposit is missed.

File: pow/src/main/java/tech/pegasys/teku/pow/ValidatingEth1EventsPublisher.java (hunk 1 above)
Comment (mbaxter) [reply pow/src/main/java/tech/pegasys/teku/pow/ValidatingEth1EventsPublisher.java]: good call - fixed

**Pull request merged**